cli/JSXV4UPGRADE.md

```diff
@@ -0,0 +1,44 @@
+**JSX V4 upgrade path**
+
+1. No upgrading rescript-react && no changes in bsconfig.json
+
+   JSX V3 by default.
+
+2. No upgrading rescript-react && new configuration added in bsconfig.json
+
+   a. JSX V3
+
+   ```json
+   "react": {
+     "jsx": 3,
+     "runtime": "classic"
+   }
+   ```
+
+   The `react.runtime` affects nothing, no matter of `"classic"` or `"automatic"`.
+
+   b. JSX V4 with classic mode
```

Comment:
This doc looks good.
One aspect to consider more is how to adopt `"jsx": 4` gradually. This `bsconfig.json` change is sufficient when the entire project is using V4. But we need to also think about what to do when half the project uses V3 and half the project uses V4.

Reply:
At the very least one needs to specify per-file `"jsx": 4` vs `"jsx": 3` i.e. override the bsconfig locally.
In fact one might need to be able to do that in the middle of a file.
E.g. What if the file is `<A > <B /> </A>` and:
1 initially both `A` and `B` are in V3, and the current file is using V3
2 make the current file use V4
3 make `A.res` use V4
4 make `B.res` use V4

Reply:
So at point 2, the jsx would cal `<A>`  the V4 way but it expects to be called the V3 way.
I think this is solved by turning on V4 at the top of the file. And locally turning on V3 where `<A > <B /> </A>` is.

Reply:
For 3, `A` uses V4 but `B` is still on V3. 
We need to be able to turn on V3 inside `<A > ... </A>` so that we can call `B` with V3.

Reply:
For 4, we don't need anything anymore. Just turn on V4 at the top of the current file.
When every file has been done like that and the project is converted, then non per-file config is needed anymore.

Does this make sense?
Is it going to work?

Reply:
If this is what we need, then the local config needs to go into an attribute. As attributes can both be used module-level (for the top of the file) or at the level of individual expressions.

Reply:
This is how it might look like. Notice the use of module-level attributes `@@` and expression-level `@`.

### 1. ###
```rescript
<A> <B/> </A>
```

### 2. ###
```rescript
@@jsxconfig(4)

...

@@jsxconfig(3)

<A> <B/> </A>
```

### 3. ###
```rescript
@@jsxconfig(4)

...

<A> @jsxconfig(3) <B/> </A>
```

### 4. ###
```rescript
@@jsxconfig(4)

...

<A> <B/> </A>
```

Reply:
This is a really good point I didn't consider enough frankly. Can you elaborate on the case that A, B are using V3 against the current file? I can imagine that A, B are all installed dependencies which has the bsconfig.json respectively. If I'm correct, I'm afraid that it is not available to use the new JSX mode in`Js.Jsx` for the first release of V4. Because A, B have types from rescript-react, e.g. `React.element`, but the current file will have a type based on `Js.Jsx.element`. Those are not mixed to compose the components as flexible as users want. I mean that the new JSX mode is not available without breaking the existing projects by combining the V3 components. IMO, the first release of V4 is available 1 ~ 2.c in my upgrade doc. Definitely we have to focus and test thoroughly the 2.b is not going to break any projects with V3 dependencies. I think we can achieve that with current implementation. Anyway, 2.c is just the experimental feature for the projects without V3 dependencies.

Reply:
And for the next stage, we can make the `Js.Jsx` matured as an experimental feature not only for the new JSX mode, so that the external dependencies are going to adopt and upgrade their modules, especially rescript-react. I think we can ship the V4 with "automatic" runtime by default later. IMHO, I'm afraid that the tweaks such as `@jsxconfig` may cause complexity issues and confusion for users.

Reply:
I think the process is:
- switch to V4 but do *not* use the new jsx mode. So everything is backwards compatible
- gradually convert the components to V4
- eventually use the new JSX mode if you like
Is that what you have in mind?

It's important that `A` and `B` be just files (or even local modules) not entire dependencies, so you can gradually switch over a large application. Sure most of the times you can just switch them over all at once, but not always.

Reply:
The majority of cases will be, just switch to V4 and there's nothing to do. But not in all cases. I think.

Reply:
I think the relevant types are things like:
```rescript
type element

type componentLike<'props, 'return> = 'props => 'return

type component<'props> = componentLike<'props, element>

/* this function exists to prepare for making `component` abstract */
external component: componentLike<'props, element> => component<'props> = "%identity"

```

In there the only "secret" type is `element`.
So in order for things to work all that is needed is that `React.element` is recognised by the type system as the same type as `Js.Jsx.element`. There are many ways one can achieve that. It's easy to define modules and open them locally to shadow type definitions.

One can define `Jsx.element` to be `React.element`. Or use `Js.React.element` defined in that way.
Or one can `open` use a module called `React` that defines `type element = Js.Jsx.element` etc etc.

Reply:
So I'd like firs to know what the primitive types are. What the primitive values are (`jsx` and `jsxDom` etc)) with their types.
Then *separately* see how to connect that to rescript-react.
And to rescript react native.
And ideally to something else too.

Reply:
So back to the `A` and `B` components, V3 vs V4 scenario.
It really depends on the nature of projects how detailed one needs to be.
The happy case is: turn on V4 and do nothing. It works.
But worries are: what if one has written some `make` functions by hand? They need to change. What if someone is exporting components to JS and then calling make directly. That needs to change too. Etc.
Since these worries are highly project-dependent, there's no way to know in advance.

Perhaps a good way to go is to document what is the worst case scenario (need to switch between V3 and V4 at arbitrary points), but don't implement it yet. Just do some tests.
E.g. take a few large apps and see what breaks, if anything.
Based on that, decide how deeply to go.

Reply:
> E.g. take a few large apps and see what breaks, if anything.

There are two apps for the good battle-test, my company apps 😏. It is quite large, I think.

Reply:
Great plan. Looking forward to hearing how it goes.

Reply:
Kind of depends on what we need. Just knowing most things are possible frees up the design space.
We don't need to go into detail with one until we know it's needed.

Proposal: how about you just write it for React. So use `React.element` etc. Then I'll have a go at the general version.

Reply:
> Proposal: how about you just write it for React. So use `React.element` etc. Then I'll have a go at the general version.

OK, I'll give it a try!

Reply:
I tried to test changing the type declaration in `React.res` of rescript-react in the node_modules.
```rescript
type element = Js.React.element
```
It seems working with composing V4 + new JSX mode & V3 in the same project. But I can't figure out how to open and shadow the type declaration after making the `React` module in my project. Technically I can make the `React.res` in my project, but It seems not possible to open `React` inside it. Is it what you meant?

<img width="489" alt="image" src="https://user-images.githubusercontent.com/36963424/176491275-36c457f0-862b-4263-b8b9-c85e22f28329.png">

Reply:
Actually I quite like Patrick's proposal of being super lightweight.
So the ppx just emits calls to `Jsx.jsx` and `Jsx.jsxDom`, and the compiler defines nothing at all.
Then RescriptReact needs to provide a module called `Jsx` instead of `React`, and that's all.

Reply:
So rescript-react can use whatever type it wants for `jsx` and `jsxDom` as long as the code generated by the ppx type checks.

Reply:
> If so, rescript-react needs to be upgraded. Because it’s binding for jsx is not right. That is all the way back to thr begining of our duscussion.

Sorry for the confusion. I guess there are too many things going at the same time and it's very easy to misunderstand.
I was under the impression that it's perfectly OK to rely on a new version of rescript-react to support the new jsx mode.
But it's essential that people can switch to V4 (not with the new jsx) without changing anything.
I thought that was the plan.

I think once people have migrated to V4, then it's not a problem to upgrade rescript-react and turn on the new jsx mode.

Reply:
Just one minor comment. It's not "they". It is "us".
As long as people are kept in the loop and questions asked in the forum when in doubt, we can just go ahead make new version of rescript-react whenever needed.

Reply:
It's you or Patrick's call, I can make PR for new `Jsx` module if you guys want. Or I'll await the `Jsx` module's interface are going to be fixed, once it is fixed, PPX V4 just emits to call that. I guess V4 is almost done now.

Reply:
Definitely a PR for the new Jsx is useful for testing. Making sure that all works.
Then make a release when the time is right.